src/coprocessor/endpoint.rs

```diff
@@ -360,54 +379,106 @@ impl TiDbEndPoint {
             on_error(e, t);
             return;
         }
-        match self.handle_select(&mut t) {
+        let resp = match t.cop_req.take().unwrap() {
+            Ok(CopRequest::Select(sel)) => self.handle_select(sel, &mut t),
+            Ok(CopRequest::DAG(dag)) => self.handle_dag(dag, &mut t),
+            Err(err) => Err(err),
+        };
+        match resp {
             Ok(r) => respond(r, t),
             Err(e) => on_error(e, t),
         }
     }
 
-    pub fn handle_select(&self, t: &mut RequestTask) -> Result<Response> {
-        let sel = match t.sel_req.take().unwrap() {
-            Ok(sel) => sel,
-            Err(err) => {
-                return Err(err);
-            }
-        };
-        let snap = SnapshotStore::new(self.snap.as_ref(),
-                                      sel.get_start_ts(),
-                                      t.req.get_context().get_isolation_level());
+    pub fn handle_select(&self, sel: SelectRequest, t: &mut RequestTask) -> Result<Response> {
+        let snap = SnapshotStore::new(self.snap.as_ref(), sel.get_start_ts(), IsolationLevel::SI);
         let mut ctx = try!(SelectContext::new(sel, snap, t.deadline, &mut t.statistics));
         let mut range = t.req.get_ranges().to_vec();
         debug!("scanning range: {:?}", range);
         if ctx.core.desc_scan {
             range.reverse();
         }
-        let res = if t.req.get_tp() == REQ_TYPE_SELECT {
-            ctx.get_rows_from_sel(range)
-        } else {
-            ctx.get_rows_from_idx(range)
+        let res = match t.req.get_tp() {
+            REQ_TYPE_SELECT => ctx.get_rows_from_sel(range),
+            REQ_TYPE_INDEX => ctx.get_rows_from_idx(range),
+            _ => unreachable!(),
```

Comment:
using match in two branch instead of using if/else is more understandable, and use unreadchable is reasonable since the bug is explicit.